irods/test/meta_test.py

```diff
@@ -29,13 +29,76 @@ def setUp(self):
+
+        with self.assertRaises(BadAVUOperationKeyword):
+            AVUOperation(operation="add", avu=iRODSMeta(*m), extra_keyword=None)
+
+
+        with self.assertRaises(InvalidAtomicAVURequest):
+            user.metadata.apply_atomic_operations( tuple() )
+
+        user.metadata.apply_atomic_operations()   # no AVUs applied - no-op without error
+
+        for n,obj in enumerate((group, user, self.coll, self.obj)):
+            avus = [ iRODSMeta('some_attribute',str(i),'some_units') for i in range(n*100,(n+1)*100) ]
+            obj.metadata.apply_atomic_operations(*[AVUOperation(operation="add", avu=avu_) for avu_ in avus])
+            obj.metadata.apply_atomic_operations(*[AVUOperation(operation="remove", avu=avu_) for avu_ in avus])
+
+
+    def test_atomic_metadata_operation_for_resource_244(self):
+        (root,leaf)=('ptX','rescX')
+        with self.create_simple_resc_hierarchy(root,leaf):
+            root_resc = self.sess.resources.get(root)   # resource objects
+            leaf_resc = self.sess.resources.get(leaf)
+            root_tuple = ('role','root','new units #1')    # AVU tuples to apply
+            leaf_tuple = ('role','leaf','new units #2')
+            root_resc.metadata.add( *root_tuple[:2] ) # first apply without units ...
+            leaf_resc.metadata.add( *leaf_tuple[:2] )
+            for resc,resc_tuple in ((root_resc, root_tuple), (leaf_resc, leaf_tuple)):
+                resc.metadata.apply_atomic_operations(  # metadata set operation (remove + add) to add units
+                    AVUOperation(operation="remove", avu=iRODSMeta(*resc_tuple[:2])),
+                    AVUOperation(operation="add", avu=iRODSMeta(*resc_tuple[:3]))
+                )
+                resc_meta = self.sess.metadata.get(Resource, resc.name)
+                avus_to_tuples = lambda avu_list: sorted([(i.name,i.value,i.units) for i in avu_list])
+                self.assertEqual(avus_to_tuples(resc_meta), avus_to_tuples([iRODSMeta(*resc_tuple)]))
+
+
+    def test_atomic_metadata_operation_for_data_object_244(self):
+        AVUs_Equal = lambda avu1,avu2,fn=(lambda x:x): fn(avu1)==fn(avu2)
+        AVU_As_Tuple = lambda avu,length=3:(avu.name,avu.value,avu.units)[:length]
+        AVU_Units_String = lambda avu:"" if not avu.units else avu.units
+        m = iRODSMeta( "attr_244","value","units")
+        self.obj.metadata.add(m)
+        meta = self.sess.metadata.get(DataObject, self.obj_path)
+        self.assertEqual(len(meta), 1)
+        self.assertTrue(AVUs_Equal(m,meta[0],AVU_As_Tuple))
+        self.obj.metadata.apply_atomic_operations(                                  # remove original AVU and replace
+           AVUOperation(operation="remove",avu=m),                                  #   with two altered versions
+           AVUOperation(operation="add",avu=iRODSMeta(m.name,m.value,"units_244")), # (one of them without units) ...
+           AVUOperation(operation="add",avu=iRODSMeta(m.name,m.value))
+        )
```

Comment:
If someone generated a list of `AVUOperation`s dynamically and wanted to pass that list to `metadata.apply_atomic_operations`, could they do that?

Can this function take an array/list as well?

Reply:
I just noticed lines 63 and 64 above. Is that how someone would achieve what I just asked about?

Reply:
Yes, those lines show the general strategy.  Distilled down, its like this:
```
def f( x, y ) : print (x, y)
def g( *z ) : print (z)
>>> f(3,4) # =>  prints (3,4)
>>> g(3,4) # => prints (3,4)
>>> b = [2,4,6,8] # or any computed list or tuple
>>> g(*b) # => prints out b, as a tuple
```

Reply:
And if you look at the function definitions for `apply_atomic_operations`, you can see they're variadic by virtue of taking a final `*parameter` where that `parameter` collects all "final" arguments being applied in the function call , however many,  into a tuple.

Reply:
i think this is covered.   good stuff.